### ISSUE TITLE
pnext: API to query table name and column name is required

### ISSUE THREAD

**Issue by bigwad** [Starcounter/Starcounter#3139]
To make replication work, I need to get table name, column name and column type based on layout_id and column index. Could you expose such API? @k-rus, @alexanderstc

**Comment by alexanderstc** [comment]
Missing e-mail notifications.

**Comment by k-rus** [comment]
Actually, this repo is not appropriate for this issue. Issues related to native meta-layer are in Level0 repo.

**Comment by bigwad** [comment]
API is not required now, since design of replicator has been changed. But "pseudo sql query" how to query metatable to extract table name and column name base on layout_id and column index is appreciated.
What I know now, that I can get table name using 
```select FullName(5) from RawView where Layout_handle(8)=?```
and for column name:
```select Name(3) from RawView, RawColumn where RawView.__id = RawColumn.Table(2) AND RawColumn.Index(8) = ?```
and if not found in RawColumn then search in ReferenceRawColumn.

Right?

**Comment by k-rus** [comment]
Metalayer does not provide query functionality. It only provides higher level API to call indexes defined on meta-data tables. They are already exported as C++ if I don't mistake. Examples of how to use them can be found in implementations of public C API.

**Comment by bigwad** [comment]
Thanks, I know that. That's why I called it pseudo query. I need them to be logically right. Physical implementation is my concern.

**Comment by k-rus** [comment]
I don't like to use SQL notation for this purpose, since using API is completely imperative.

Since you want to retrieve column by `LayoutHandle` and `ColumnIndex`, you need to find the record in `RawView` for `LayoutHandle`. Then you can use compound the primary key index on `RawColumn` to retrieve a column and then its name. You don't need to search `ReferenceRawColumn`, since `RawColumn` extent contains children too, i.e., records in `ReferenceRawColumn`.

**Comment by bigwad** [comment]
@k-rus, [here](https://github.com/Starcounter/CompanyTrack/issues/137#issuecomment-158057457) in pp 11 and 16 you make some statements about types and layouts. If summarize - different types may refer to the same layout, but not to the same setspecifier.
I want to unambiguously find typename base on layout. Obviously, it's not possible if above statement is true.
So the questions are:
- is it actually implemented that several types may refer to the same layout?
- if yes - what is the algorithm to find typename using layout and maybe some other data that is available from the record itself (like setspecifer)?

**Comment by k-rus** [comment]
> is it actually implemented that several types may refer to the same layout?

No, it is not implemented so, but it will be changed in future.
It will be needed, e.g., for dynamic types. So the assumption is that the same layout can be used in objects of different types.



> if yes - what is the algorithm to find typename using layout and maybe some other data that is available from the record itself (like setspecifer)?

Normally, you don't search type by layout. Each record knows its type from set specifier. Each type has an expected layout, which is used for creating records. This information is stored in meta-data. For different reasons, several types will have the same layout as expected layout.

**Comment by bigwad** [comment]
>No, it is not implemented so, but it will be changed in future.

What kind of "future" do you mean? Is it going to be changed in upcoming release or it's a far future plans?

>Normally, you don't search type by layout. 

My task is not normal. So I assume, that now layout is enough, but in the future setspecifier will be required. The crucial question - when will this future come?

**Comment by k-rus** [comment]
> The crucial question - when will this future come?

I don't plan future. I guess there is few months before.
If you understand that your current solution is temporal and you plan to work on a proper solution soon, then it is probably acceptable to implement current work around using layout.
I just want to make sure we don't run in trouble in future
